src/ucx_van.h

```diff
@@ -451,256 +553,356 @@ class UCXVan : public Van {
     ucs_status_t status     = ucp_listener_create(worker_, &params, &listener_);
     CHECK_STATUS(status) << "ucp_listener_create failed: " << ucs_status_string(status);
 
-    UCX_LOG(1, "bound to " << addr.sin_addr.s_addr << " port: " << node.port);
+    UCX_LOGE(1, "bound to " << addr.sin_addr.s_addr << " port: " << port);
+  }
 
-    return node.port;
+  void Connect(const Node &node) {
+      ep_pool_.Create(node);
   }
 
-  void Connect(const Node &node) override {
-    CHECK_NE(node.id, node.kEmpty);
-    CHECK_NE(node.port, node.kEmpty);
-    CHECK(node.hostname.size());
+  void Poll() {
+    ucp_tag_message_h msg;
+    ucp_tag_recv_info_t info;
+    int cnt = 0;
 
-    // worker doesn't need to connect to the other workers. same for server
-    if ((node.role == my_node_.role) && (node.id != my_node_.id)) {
+    // Poll all underlying transports(IB, shm, tcp, etc)
+    cnt = ucp_worker_progress(worker_);
+    if (cnt == 0) {
       return;
     }
 
-    ep_pool_.Create(node);
+    // Check whether any meta data or push/pull request arrived
+    do {
+      // only match the highest bit of tag.
+      msg = ucp_tag_probe_nb(worker_,
+                             static_cast<ucp_tag_t>(Tags::UCX_TAG_META),
+                             static_cast<ucp_tag_t>(Tags::UCX_TAG_MASK),
+                             1, &info);
+      if (msg != NULL) {
+        // some meta data is ready, post a receive to get it
+        UCXRequest *meta_req = PostRecvMeta(msg, &info);
+        if (meta_req->completed) {
+          // meta data received immediately, can post receive for a real
+          // data, because it's length is known from meta
+          PostRecvData(meta_req);
+        }
+      }
+    } while (msg != NULL);
   }
 
-  int SendMsg(Message &msg) override {
-    int id = msg.meta.recver;
-    CHECK_NE(id, Meta::kEmpty);
+  ucs_status_ptr_t Send(Message &msg, void *buf, size_t count, ucp_datatype_t dt,
+                        Tags tag) {
+    int id         = msg.meta.recver;
+    int dst_dev_id = msg.meta.dst_dev_id;
+    ucp_ep_h ep    = ep_pool_.Find(id, dst_dev_id);
+    ucp_tag_t stag = MakeTag(my_node_->id, tag, msg.meta.key);
 
-    ucp_ep_h ep = ep_pool_.Find(id);
-    if (!ep) {
-      LOG(ERROR) << "there is no UCX endpoint to node " << id;
-      return -1;
-    }
+    UCX_LOGE(1, "Send to ep " << ep  << " (" << id << ", " << dst_dev_id);
```

Comment:
missing a closing `)` in the log message

Reply:
fixed